### PR TITLE
mgr/volumes: unregister job upon async threads exception

### PR DESCRIPTION
**Pull request by vshankar**

### CODE REVIEW

**Comment by ajarr** [comment]
So I guess for nautilus we'd cherry pick this fix and fix the py2 issue mentioned here,
https://tracker.ceph.com/issues/44293#note-2
Venky, can also include Fixes:https://tracker.ceph.com/issues/44293 tag in the commit message.

**Comment by vshankar** [comment]
> So I guess for nautilus we'd cherry pick this fix and fix the py2 issue mentioned here,
> https://tracker.ceph.com/issues/44293#note-2
> Venky, can also include Fixes:https://tracker.ceph.com/issues/44293 tag in the commit message.

Looking again at the backtrace -- not sure if it's py2 related, but, it has to do something with escaping strings.

The tracker ticket is for nautilus -- we need one for upstream, correct?

**Comment by vshankar** [comment]
@batrick @ajarr -- ran nautilus branch with this fix in teuthology: http://pulpito.ceph.com/vshankar-2020-02-26_09:27:32-fs-wip-nautilus-mgr-volumes-testing-basic-smithi/

All failures are https://tracker.ceph.com/issues/44276

File: src/pybind/mgr/volumes/fs/async_job.py [hunk 1]
```diff
@@ -56,7 +53,12 @@ def run(self):
                 exc_type, exc_value, exc_traceback = sys.exc_info()
                 log.warning("traceback: {0}".format("".join(
                     traceback.format_exception(exc_type, exc_value, exc_traceback))))
-            time.sleep(1)
+            finally:
+                # when done, unregister the job
+                if vol_job:
+                    with self.async_job.lock:
+                        self.async_job.unregister_async_job(vol_job[0], vol_job[1], thread_id)
+                time.sleep(1)
```
Comment (batrick) [inline src/pybind/mgr/volumes/fs/async_job.py]: Why move the sleep into the `finally` clause?

**Comment by batrick** [comment]
> > So I guess for nautilus we'd cherry pick this fix and fix the py2 issue mentioned here,
> > https://tracker.ceph.com/issues/44293#note-2
> > Venky, can also include Fixes:https://tracker.ceph.com/issues/44293 tag in the commit message.
> 
> Looking again at the backtrace -- not sure if it's py2 related, but, it has to do something with escaping strings.
> 
> The tracker ticket is for nautilus -- we need one for upstream, correct?

https://tracker.ceph.com/issues/44293#note-3

File: src/pybind/mgr/volumes/fs/async_job.py (hunk 1 above)
Comment (vshankar) [reply src/pybind/mgr/volumes/fs/async_job.py]: doesn't make much difference anyway, right? is the concern about code readability?

Comment (batrick) [reply src/pybind/mgr/volumes/fs/async_job.py]: Just stuck out to me.

**Comment by batrick** [comment]
@vshankar in any case, please amend the commit with a `Fixes:` line.

**Comment by batrick** [comment]
@vshankar https://tracker.ceph.com/issues/44315

Maybe we can get this merged before v14.2.8 is complete. Please do the backport and post a new round of tests.

**Comment by vshankar** [comment]
> Maybe we can get this merged before v14.2.8 is complete. Please do the backport and post a new round of tests.

ack

**Pull request merged**